rocAL_pybind/amd/rocal/plugin/jax.py

```diff
@@ -0,0 +1,295 @@
+
+        if sharding is not None:
+            assert isinstance(
+                sharding, (NamedSharding, PositionalSharding)
+            ), "`sharding` should be an instance of `NamedSharding` or `PositionalSharding`"
+        self._sharding = sharding
+        self._is_data_available = False
+
+    def next(self):
+        return self.__next__()
+
+    def __next__(self):
+        self._is_data_available = True
+        pipeline_outputs = []
+        for self.loader in self.pipelines:
+            if self.loader.rocal_run() != 0:
+                raise StopIteration
+            self.output_tensor_list = self.loader.get_output_tensors()
+
+            self.output_list = []
+            self.device_id = self.loader._device_id
+            if self.loader._name is None:
+                self.loader._name = self.loader._reader
+            self.last_batch_policy = self.loader._last_batch_policy
+            assert (
+                self.last_batch_policy != types.LAST_BATCH_PARTIAL
+            ), "JAX iterator does not support partial last batch policy."
+            self.labels_size = ((self.batch_size * self.loader._num_classes)
+                                if self.loader._one_hot_encoding else self.batch_size)
+            for i in range(len(self.output_tensor_list)):
+                self.dimensions = self.output_tensor_list[i].dimensions()
+                self.dtype = self.output_tensor_list[i].dtype()
+                self.output = convert_to_jax_array(
+                    self.output_tensor_list[i].__dlpack__(self.device_id))
+                self.output_list.append(self.output)
+
+            if self.loader._name == "labelReader":
+                if self.loader._one_hot_encoding == True:
+                    self.labels = np.empty(self.labels_size, dtype="int32")
+                    self.loader.get_one_hot_encoded_labels(
+                        self.labels.ctypes.data, self.loader._output_memory_type)
+                    self.labels_tensor = self.labels.reshape(
+                        -1, self.batch_size, self.loader._num_classes)
+                    self.labels_tensor = convert_to_jax_array(
+                        self.labels_tensor)
+                else:
+                    self.labels = self.loader.get_image_labels()
+                    self.labels_tensor = self.labels.astype(dtype=np.int_)
+                    self.labels_tensor = convert_to_jax_array(
+                        self.labels_tensor)
+                self.labels_tensor = jax.device_put(
+                    self.labels_tensor, self.output_list[0].device)
+                self.output_list.append(self.labels_tensor)
+            pipeline_outputs.append(self.output_list)
+
+        if self.num_devices == 1 and self._sharding is None:
+            return pipeline_outputs[0]
+
+        sharded_outputs = []
+        for i in range(len(pipeline_outputs[0])):
```

Comment:
What is len(pipeline_outputs[0]) is it for the batch size?

Reply:
Each pipeline provides a list of outputs which then gets stored into pipeline_outputs list. We take the length of pipeline_outputs[0] to iterate through all the outputs in each pipeline.